### ISSUE TITLE
Outdated version 0.2.0 on Julia 1.6.1

### ISSUE THREAD

**Issue by nxtruong** [MilesCranmer/SymbolicRegression.jl#24]
I am using Julia 1.6.1. When I added SymbolicRegression, the version was very much outdated: 0.2.0. The first example in the docs didn't run because EquationSearch didn't exist. Can you fix the dependency so that the latest version supports Julia 1.6.1?

**Comment by MilesCranmer** [comment]
Thanks; I will look into this a bit more. I have the CI testing run on 1.6 [here](https://github.com/MilesCranmer/SymbolicRegression.jl/blob/master/.github/workflows/CI.yml), but it seems this same error is coming up for a few people. Let me see what I can do.

**Comment by MilesCranmer** [comment]
I can't reproduce this. The SymbolicRegression.jl version downloaded for 1.6.1 is up-to-date for me. What package server are you pulling from?

Also related to #22. I think these issues are due to pulling an out-of-date package.


As a quick fix, do the following:

Delete `~/.julia/registries/General`, then run the following commands in Julia:

```
ENV["JULIA_PKG_SERVER"] = ""
import Pkg
Pkg.update()
```

**Comment by nxtruong** [comment]
I was able to install the latest version by following your instructions. Thanks. It works now.

**Comment by nxtruong** [comment]
I have to re-open this issue because I still have it. When I followed the instructions above in a new environment, the installed version of `SymbolicRegression` was the latest. However, in the environment of one of my projects, even if I followed the instructions above, I still only got the version 0.2.0 of `SymbolicRegression`. I think there must be some kind of dependency conflicts in the installed packages, so ultimately it might not be an issue with `SymbolicRegression` but I don't know how to find out why and how to fix the issue. The status of the environment where I had the issue is given below (you can see the old version of `SymbolicRegression`).
```
  [336ed68f] CSV v0.8.4
  [f65535da] Convex v0.14.8
  [2445eb08] DataDrivenDiffEq v0.5.7
  [a93c6f00] DataFrames v1.1.0
  [28b8d3ca] GR v0.57.4
  [2e9cd046] Gurobi v0.9.12
  [682c06a0] JSON v0.21.1
  [961ee093] ModelingToolkit v5.16.0
  [91a5bcdd] Plots v1.13.2
  [438e738f] PyCall v1.92.3
  [ebc72ef8] SciPy v0.1.1
  [8254be44] SymbolicRegression v0.2.0
  [0c5d862f] Symbolics v0.1.25
  [1986cc42] Unitful v1.7.0
```

**Comment by MilesCranmer** [comment]
Hm, this is strange. Hopefully I can help solve it.
It seems like you might be installing SymbolicRegression into a different environment than the one you are using?

Can you launch julia with `julia --project=my/julia/env`, where `my/julia/env` is the directory containing the `Project.toml` file describing your environment? Then, you can do `import Pkg; Pkg.update()` to update packages. 

Further, you can fix the version in `Project.toml` by adding:
```
[compat]
SymbolicRegression = "0.5.16"
```
at the end of the file.

My guess is the environment you are running from is different than the environment you are installing to.

If this still gives you problems, you could try the Python frontend: [PySR](https://github.com/MilesCranmer/PySR), which does the environment setup for you automatically.

**Comment by nxtruong** [comment]
I know how to work with environments. I tried to figure out which package has version / dependency conflicts with `SymbolicRegression`. So I created a new, empty environment; added `SymbolicRegression` (version 0.5.16), then gradually added the other packages listed above, one by one. It turns out that the package `Symbolics` ([link](https://github.com/JuliaSymbolics/Symbolics.jl)) caused `SymbolicRegression` to downgrade to version 0.2.0 as shown above. So, apparently, there is some conflict between `SymbolicRegression` and `Symbolics`.

I also noticed that if I forced `SymbolicRegression` to be at version 0.5.16 (by `add SymbolicRegression@0.5.16` not by editing the `Project.toml` file), then `SymbolicRegression` would be at the correct version but it forced several other packages to be downgraded, for example: `ModelingToolkit` down to 4.5.0 from 5.17.0, `SymbolicUtils` v0.11.2 ⇒ v0.6.3, `LoopVectorization` v0.11.2 ⇒ v0.10.0, and others. Perhaps the `Project.toml` file of `SymbolicRegression` can be updated to include more recent versions of its dependencies?

**Comment by MilesCranmer** [comment]
Thanks for taking the time to figure this out! Also, thanks for documenting this; this info is very helpful for me. This probably explains this issue that others have had as well.

So it looks like I will need to update the compatibility to the new Symbolics libraries. I was not aware that using older versions of Julia packages would result in this type of issue where Pkg will use such an old version of SymbolicRegression. SymbolicUtils has had several core API and type changes recently that will require a fair amount of rewriting - this is why I haven't implemented #21 / #20 yet. But it looks like I will need to get on this.

Thanks! I'll let you know when this is fixed. Until then, I would recommend keeping a separate `Project.toml` and fixing the SymbolicRegression version in it. Unfortunately, as you observed, this means the other package versions in the env will be downgraded.
Best regards,
Miles

**Comment by MilesCranmer** [comment]
Quick update: I tried to get this working, but the entire SymbolicUtils interface has changed down to the core types themselves.  e.g., an expression which used to be of type `SymbolicUtils.Term{Number}` is now the type `SymbolicUtils.Add{Number, Float64, Dict{Any, Number}, Nothing}` - and similar for multiplication and power, but not for user-defined operators. This makes things very complex to adapt to.

So, if I do want to update SymbolicUtils, I think it will be much easier for me to completely write my simplification backend from scratch, rather than update my calls to their new types, since it is very structurally different. I would want to move away from relying on these type definitions, and just use their [interface](https://symbolicutils.juliasymbolics.org/interface/). It will probably be a bit slower (that's why I used their internal types originally), but I think compatibility with their package is a good thing.

But this is a big change and I don't have a timeline on it. So for now, one unfortunately needs to fix the version of SymbolicRegression.jl in Project.toml.

Cheers,
Miles
